packages/dd-trace/src/collector/span.js

```diff
@@ -0,0 +1,118 @@
+'use strict'
+
+const { channel } = require('dc-polyfill')
+const DatadogSpan = require('../opentracing/span')
+const DatadogCollectorSpanContext = require('./span_context')
+
+const now = performance.now.bind(performance)
+
+const startSegmentChannel = channel('datadog:tracing:segment:start')
+const segmentDiscardChannel = channel('datadog:tracing:segment:discard')
+const startChannel = channel('datadog:tracing:span:start')
+const tagsChannel = channel('datadog:tracing:span:tags')
+const errorChannel = channel('datadog:tracing:span:error')
+const finishChannel = channel('datadog:tracing:span:finish')
+
+let segmentId = 0
+
+class DatadogCollectorSpan extends DatadogSpan {
+  finish (finishTime) {
+    if (this._spanContext._isFinished) return
+
+    this._spanContext._isFinished = true
+    this._spanContext._trace.active--
+
+    const trace = this._spanContext._trace
+
+    // TODO: Emit a discard event from tracer. For now we just short-circuit.
+    if (trace.isDiscarded) return
+    if (trace.isRecording === false) {
+      trace.isDiscarded = true
+      return segmentDiscardChannel.publish({ segmentId })
+    }
+
+    const ticks = finishTime
+      ? now() + finishTime - trace.ticks - trace.startTime
+      : now() - trace.ticks
```

Comment:
Can we do this on the collector side? Otherwise we need to keep the timing state in the library. I feel like it _should_ be safe to just send a Date.now() in the start event and monotonic timestamps in each of the start and end, then compute the delta in the collector.

Reply:
That's basically what this is doing except that it does that at the trace level instead of the span level. This is for efficiency but also because it was otherwise causing intermittent slight drifts. It's also worth noting that the computation is only happening when a finish time if provided, otherwise it only sends the monotonic timestamp.

Reply:
In both cases though you're holding the `trace.ticks` value. I'm suggesting we should try to avoid holding that in the library if at all possible. And I'm not sure why there would be drift coming from the same process on the same machine between monotonic timestamps that both came from that same process. 🤔

Reply:
This is also an optimization so that for applications under high load that generate traces quickly the events only need to send <=32 bits for the ticks instead of constantly 64 bits.

It also means an easier to reason about floor for the ticks which would always be 0 for the segment. However, since the number is arbitrary anyway, that doesn't actually matter too much.

Basically, it seemed to me that adding a single subtraction operation didn't add enough complexity to justify going with the alternative that saves a subtraction but ends up sending a lot more data. With the current structure of events, the timestamp is basically otherwise the largest piece of data of an event. I even considered going with microseconds since nanoseconds are rarely (read: never) useful for tracing data.

Reply:
But you're trading an extra potential maximum of 32 bits of data over the wire for _more than that_ held in process memory for the duration of that span. That adds up quickly when there are potentially hundreds of thousands of concurrent spans. Process memory is a lot more scarce than bandwidth.

More concerning to me though is that we need a place to _store_ any data like this which is held in-process for the span duration. I'm _trying_ to get us _away_ from storing the big bag of data that is SpanContext, but things like this further entangle it.

Reply:
It's actually the opposite. It trades the memory that would _otherwise be held in memory_ for every span because the data is buffered, to a single integer that is held only once for the whole segment. So this actually _saves_ memory in-process, and also has the benefit of reducing bandwidth. You could argue it increases CPU usage a tiny bit by adding a subtraction, but that's minimal enough to justify the trade-off.

Reply:
I still don't understand where the problem is. Correct me if I'm wrong, but these seem to be the 2 approaches we are discussing:

Option 1:
- Send timestamp and ticks for segment.
- Send raw ticks for the span.

Option 2:
- Send timestamp for segment.
- Send computed ticks for span.

For option 1, you end up having to always send 64bit values. For option 2, you often need less than 64bit since the values will start from 0 instead of the ticks when the segment started. So option 2 uses less memory, at the cost of slightly more CPU for a subtraction per span, but that is an extremely low cost, and would be offset anyway by the need to write less data to the buffer.

I don't understand how putting _less data_ in the buffer is supposed to result in _more memory_? Can you clarify your calculation?

Reply:
Computed ticks requires holding additional data on the Span Context between all the captured points in the segment, while just using raw ticks everywhere does not. It also makes it more complicated to take a more stream-oriented approach with less segment buffering as the timestamp the computed ticks are in relation to changes and would therefore add additional complexity to translating them back to absolute timestamps for backend processing and frontend visualization.

There's not a whole lot of benefit to being event-based if we're going to just continue buffering all the data in-process anyway. We're just moving held memory around rather than actually getting it _out_ of the process which is the whole _point_ of wanting to go for an event model.